overlord/state/task.go

```diff
@@ -196,26 +206,59 @@ func (t *Task) Status() Status {
 	return t.status
 }
 
+func (t *Task) changeStatus(old, new Status) {
+	t.status = new
+	if !old.Ready() && new.Ready() {
+		t.readyTime = timeNow()
+	}
+	chg := t.Change()
+	if chg != nil {
+		chg.taskStatusChanged(t, old, new)
+	}
+}
+
 // SetStatus sets the task status, overriding the default behavior (see Status method).
 func (t *Task) SetStatus(new Status) {
+	if new == WaitStatus {
+		// Disallow going into WaitStatus, the Wait() API should be used instead.
+		return
+	}
+
 	t.state.writing()
 	old := t.status
-	if (new == DoneStatus || new == WaitStatus) && old == AbortStatus {
+	if new == DoneStatus && old == AbortStatus {
 		// if the task is in AbortStatus (because some other task ran
 		// in parallel and had an error so the change is aborted) and
-		// DoneStatus/WaitStatus was requested (which can happen if the
+		// DoneStatus was requested (which can happen if the
 		// task handler sets its status explicitly) then keep it at
 		// aborted so it can transition to Undo.
 		return
 	}
-	t.status = new
-	if !old.Ready() && new.Ready() {
-		t.readyTime = timeNow()
-	}
-	chg := t.Change()
-	if chg != nil {
-		chg.taskStatusChanged(t, old, new)
+	t.changeStatus(old, new)
+}
+
+// Wait puts the task into WaitStatus, and sets the status the task should be restored
+// to after the Wait.
+func (t *Task) Wait(resultStatus Status) {
+	t.state.writing()
+	old := t.status
+	if old == AbortStatus {
+		// if the task is in AbortStatus (because some other task ran
+		// in parallel and had an error so the change is aborted) and
+		// WaitStatus was requested (which can happen if the
+		// task handler sets its status explicitly) then keep it at
+		// aborted so it can transition to Undo.
+		return
 	}
+	t.waitStatus = resultStatus
+	t.changeStatus(old, WaitStatus)
+}
+
+// WaitStatus returns the status the Task should return to once the current WaitStatus
+// has been resolved.
+func (t *Task) WaitStatus() Status {
```

Comment:
Naming check please, it conflicts with state.WaitStatus, but I'm out of ideas for improvement

Reply:
I think WaitedStatus or AfterWaitStatus might be clearer names for this, same for the fields